### ISSUE TITLE
Library instrumentation for java.util.concurrent (metrics)

### ISSUE THREAD

**Issue by ZawadaZSE** [open-telemetry/opentelemetry-java-instrumentation#6201]
**Is your feature request related to a problem? Please describe.**
We're migrating from opentracing and noticed that manual instrumentation for Executors is missing.

**Describe the solution you'd like**
I'd like to know if you plan to implement similar solution to: https://github.com/opentracing-contrib/java-concurrent that would allow manual instrumentation of Executors.

**Describe alternatives you've considered**
I know that it's supported with javaagent instrumentation but we'd rather not use it.

**Additional context**

**Comment by mateuszrzeszutek** [comment]
Hey @ZawadaZSE ,
Do you need to create spans for every submitted task? Or is it just about context propagation across threads? If it's the latter, there already are facilities to help you with that: you can use the [`Context.taskWrapping()`](https://github.com/open-telemetry/opentelemetry-java/blob/main/context/src/main/java/io/opentelemetry/context/Context.java#L137) method to decorate an `ExecutorService`/`Executor` with current context propagation.

**Comment by reactivejson** [comment]
Hey @mateuszrzeszutek 
We are using [Context.taskWrapping()](https://github.com/open-telemetry/opentelemetry-java/blob/main/context/src/main/java/io/opentelemetry/context/Context.java#L137) method to decorate an ExecutorService.
But we are getting errors since we can't use io.opentelemetry.context.CurrentContextExecutorService with Micrometer  ExecutorServiceMetrics:
i.m.c.i.b.jvm.ExecutorServiceMetrics - Failed to bind as io.opentelemetry.context.CurrentContextExecutorService is unsupported.

**Comment by mateuszrzeszutek** [comment]
Hey @reactivejson ,
Can you try using micrometer first, and then wrapping with `Context.taskWrapping()`?
E.g.
```java
Context.taskWrapping(
    ExecutorServiceMetrics.monitor(meterRegistry, executorService, "myExecutor"))
```

**Comment by ZawadaZSE** [comment]
@mateuszrzeszutek 
Yes, we're using spring-actuator and we're manually configuring ExecutorServiceMetrics beans like this:
```
    @Bean
    ExecutorServiceMetrics someExecutorServiceMetrics(ExecutorService someExecutorService) {
        return new ExecutorServiceMetrics(someExecutorService, "some-executor", emptyList());
    }
```

As for open telemetry, we're using spring starter but that one does not provide auto configuration for executors.
We wrote our own BeanPostProcessor which wraps all ExecutorService type beans in `Context.taskWrapping(executorService)`

**Comment by mateuszrzeszutek** [comment]
You can make use of the `ExecutorServiceMetrics.monitor()` function, which also returns an `ExecutorService`, and redefine the `someExecutorService` bean the following way:

```java
@Bean 
ExecutorService someExecutorService(MeterRegistry meterRegistry) {
  ExecutorService executor = new ThreadPoolExecutor(...);
  return ExecutorServiceMetrics.monitor(meterRegistry, executor, "some-executor");
}
```

That way your `BeanPostProcessor` will wrap the already metricized executor.

I agree with you that the OpenTelemetry Spring starter should offer context propagation by default, and that it should play well with the Micrometer metrics installed by the Spring Actuator (that only measure [Spring task executors](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/task/TaskExecutorMetricsAutoConfiguration.java)), but I don't think that we can avoid wrapping `Executor`s/`ExecutorService`s -- in the first place, the OpenTracing lib you mentioned does exactly that, it wraps standard Java executors, and has separate extending wrapper classes for spring types.
